File: src/main/java/fi/aalto/cs/apluscourses/utils/PluginAutoInstaller.java
package fi.aalto.cs.apluscourses.utils;

import com.intellij.ide.plugins.PluginEnabler;
import com.intellij.ide.plugins.PluginManagerCore;
import com.intellij.ide.plugins.PluginManagerMain;
import com.intellij.ide.plugins.PluginNode;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.extensions.PluginId;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.ThrowableComputable;
import com.intellij.util.concurrency.annotations.RequiresEdt;
import fi.aalto.cs.apluscourses.intellij.notifications.NetworkErrorNotification;
import fi.aalto.cs.apluscourses.intellij.notifications.Notifier;
import fi.aalto.cs.apluscourses.ui.utils.PluginInstallerCallback;
import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class PluginAutoInstaller {

  private static class PluginInstallerWorker implements ThrowableComputable<Boolean, RuntimeException> {

    private final @Nullable Project project;

    private final @Nullable Notifier notifier;

    private final @NotNull List<PluginNode> downloadablePluginNodes;

    private PluginInstallerWorker(@Nullable Project project,
                                  @Nullable Notifier notifier,
                                  @NotNull List<PluginNode> downloadablePluginNodes) {
      this.project = project;
      this.notifier = notifier;
      this.downloadablePluginNodes = downloadablePluginNodes;
    }

    @Override
    public Boolean compute() {
      final var indicator = Objects.requireNonNull(ProgressManager.getInstance().getProgressIndicator());
      indicator.setIndeterminate(true);

      // The semaphore must be released (i.e. permits must be raised) once per each plugin installed.
      // Only when the semaphore's permit count reaches 1, this function can return.
      final Semaphore lock = new Semaphore(-downloadablePluginNodes.size() + 1);

      // "true" means that installation is successful (so far)
      AtomicBoolean installationResult = new AtomicBoolean(true);

      ApplicationManager.getApplication().invokeLater(() -> {
        try {
          final Consumer<Boolean> onInstallationComplete = (result) -> {
            installationResult.set(installationResult.get() & result);
            lock.release();
          };

          WriteAction.run(() -> PluginManagerMain.downloadPlugins(downloadablePluginNodes, Collections.emptyList(),
              false, null, PluginEnabler.getInstance(), ModalityState.defaultModalityState(), onInstallationComplete));
        } catch (IOException ex) {
          if (notifier != null) {
            notifier.notify(new NetworkErrorNotification(ex), project);
          }

          installationResult.set(false);
        }
      });

      for (; ; ) {
        // Semaphore acquisition will succeed once all plugins have been installed.
        if (lock.tryAcquire()) {
          // If plugin installation failed at any point (for example: an attempt to install a non-existent plugin),
          // we return null to indicate a failed operation.
          return installationResult.get() ? false : null;
        }

        // This will throw an exception if a user has cancelled the operation. It won't actually cancel
        // the installation of the current plugin; that would corrupt the IDE. But it will at least return
        // from this function and close the progress bar.
        indicator.checkCanceled();

        // In practice, an interruption should never happen.
        try {
          Thread.sleep(50);
        } catch (InterruptedException ex) {
          Thread.currentThread().interrupt();
          return null;
        }
      }
    }
  }

  private static final Logger logger = APlusLogger.logger;

  private static boolean shouldDownloadPlugin(@NotNull PluginId id) {
    return !PluginManagerCore.isPluginInstalled(id);
  }

  private static boolean shouldEnablePlugin(@NotNull PluginId id) {
    final var plugin = PluginManagerCore.getPlugin(id);
    return plugin != null && !plugin.isEnabled();
  }

  /**
   * Verifies that all plugins with given plugin IDs are installed and enabled. If not, the function
   * asks the user for consent to install and enable all missing plugins.
   *
   * @param project     The project which initiated the operation.
   * @param notifier    A notifier instance, used to show a notification if something goes wrong.
   * @param pluginNames List of plugin ID strings. The plugin ID can be found on the plugin's
   *                    JetBrains Marketplace website.
   * @param callback    An interface implementing functions invoked by this method. The functions
   *                    are related to showing the consent dialog box to the user.
   * @return True, if all dependencies are satisfied and there's nothing to be done. False, if
   * some dependencies were missing, and they were installed or enabled. Null, if the operation
   * has been cancelled due to missing consent or an error.
   */
  @RequiresEdt
  public static Boolean ensureDependenciesInstalled(@Nullable Project project,
                                                    @Nullable Notifier notifier,
                                                    @NotNull List<PluginDependency> pluginNames,
                                                    @NotNull PluginInstallerCallback callback) {
    final var pluginIDs = pluginNames.stream().map(p -> PluginId.getId(p.getId())).collect(Collectors.toSet());

    // Downloading and enabling plugins are different operations, so we handle these separately.
    final var pluginsToDownload = pluginIDs.stream()
        .filter(PluginAutoInstaller::shouldDownloadPlugin).collect(Collectors.toSet());
    final var pluginsToEnable = pluginIDs.stream()
        .filter(PluginAutoInstaller::shouldEnablePlugin).collect(Collectors.toSet());

    pluginsToDownload.forEach(id -> logger.info("Plugin to download: %s".formatted(id.getIdString())));
    pluginsToEnable.forEach(id -> logger.info("Plugin to enable: %s".formatted(id.getIdString())));

    if (pluginsToDownload.isEmpty() && pluginsToEnable.isEmpty()) {
      return true;
    }

    // allPluginNames = list of PluginDependencies for all plugins in pluginsToDownload or pluginsToEnable
    final var missingPluginIds = Stream.concat(pluginsToDownload.stream(), pluginsToEnable.stream())
        .map(PluginId::getIdString).collect(Collectors.toSet());
    final var allPluginNames = pluginNames.stream()
        .filter(p -> missingPluginIds.contains(p.getId()))
        .collect(Collectors.toList());

    final var consentResult = callback.askForInstallationConsent(allPluginNames);
    if (consentResult == PluginInstallerCallback.ConsentResult.IGNORE_INSTALL) {
      return true; // pretend that everything went fine
    } else if (consentResult == PluginInstallerCallback.ConsentResult.REJECTED) {
      return null; // signal to the caller to abort the operation
    }

    // Enable all requires plugins that are installed, but disabled.
    PluginEnabler.getInstance().enableById(pluginsToEnable);

    final var downloadablePluginNodes = pluginsToDownload.stream()
        .map(PluginNode::new).collect(Collectors.toList());

    final var installerWorker = new PluginInstallerWorker(project, notifier, downloadablePluginNodes);

    return ProgressManager.getInstance().runProcessWithProgressSynchronously(installerWorker,
        "Downloading required plugins", true, project);
  }

  private PluginAutoInstaller() {

  }
}

File: src/main/kotlin/fi/aalto/cs/apluscourses/utils/PluginAutoInstaller.kt
package fi.aalto.cs.apluscourses.utils

import com.intellij.ide.plugins.PluginManagerCore.isDisabled
import com.intellij.ide.plugins.PluginManagerCore.isPluginInstalled
import com.intellij.ide.plugins.PluginNode
import com.intellij.openapi.application.EDT
import com.intellij.openapi.components.service
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.util.application
import fi.aalto.cs.apluscourses.MyBundle.message
import fi.aalto.cs.apluscourses.api.CourseConfig.RequiredPlugin
import fi.aalto.cs.apluscourses.notifications.DisabledPluginsNotification
import fi.aalto.cs.apluscourses.services.course.CourseSetupStatus
import fi.aalto.cs.apluscourses.services.course.SetupAction
import fi.aalto.cs.apluscourses.services.course.SetupStepState
import fi.aalto.cs.apluscourses.services.Notifier
import fi.aalto.cs.apluscourses.services.PluginSettings
import fi.aalto.cs.apluscourses.services.Plugins
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object PluginAutoInstaller {
    internal fun shouldDownloadPlugin(id: PluginId): Boolean {
        return !isPluginInstalled(id)
    }

    internal fun shouldEnablePlugin(id: PluginId): Boolean {
        return isPluginInstalled(id) && isDisabled(id)
    }

    /**
     * Makes sure the plugins the course requires are installed and enabled.
     *
     * Missing plugins are downloaded, which takes effect after a restart. Plugins that are
     * installed but disabled are only reported, enabling one programmatically requires internal
     * platform API, so [DisabledPluginsNotification] points the user at the plugin manager instead.
     *
     * @param project     The project which initiated the operation.
     * @param pluginNames Plugins the course requires. The plugin ID can be found on the plugin's
     * JetBrains Marketplace website.
     * @return False when plugins were downloaded and the IDE has to restart for them to load.
     * True in all other cases.
     */
    suspend fun ensureDependenciesInstalled(
        project: Project,
        pluginNames: List<RequiredPlugin>,
        askForConsent: Boolean = false
    ): Boolean {
        val pluginIDs = pluginNames.map { PluginId.getId(it.id) }.toSet()
        // Downloading and enabling plugins are different operations, so we handle these separately.
        val pluginsToDownload = pluginIDs.filter { shouldDownloadPlugin(it) }.toSet()
        val disabledPlugins = pluginIDs.filter { shouldEnablePlugin(it) }.toSet()

        pluginsToDownload.forEach { CoursesLogger.info("Plugin to download: ${it.idString}") }
        disabledPlugins.forEach { CoursesLogger.info("Plugin is disabled: ${it.idString}") }

        registerRequiredPlugins(project, pluginNames)

        val setup = CourseSetupStatus.getInstance(project)
        if (disabledPlugins.isNotEmpty()) {
            val names = displayNamesOf(disabledPlugins, pluginNames)
            Notifier.notify(DisabledPluginsNotification(names, project), project)
            setup.report(
                CourseSetupStatus.PLUGINS,
                names.joinToString(", "),
                SetupStepState.ACTION_NEEDED,
                message("ui.setup.pluginsDisabled"),
                SetupAction.OPEN_PLUGIN_SETTINGS
            )
        }

        if (pluginsToDownload.isEmpty()) {
            if (disabledPlugins.isEmpty()) {
                setup.update(CourseSetupStatus.PLUGINS, SetupStepState.DONE)
            }
            return true
        }

        if (askForConsent) {
            val install = withContext(Dispatchers.EDT) {
                Messages.showOkCancelDialog(
                    project,
                    message(
                        "ui.pluginInstallationDialog.courseOpenDialog.message",
                        displayNamesOf(pluginsToDownload, pluginNames).joinToString(", ")
                    ),
                    message("ui.pluginInstallationDialog.courseOpenDialog.title"),
                    message("ui.pluginInstallationDialog.courseOpenDialog.yesText"),
                    message("ui.pluginInstallationDialog.courseOpenDialog.noText"),
                    Messages.getQuestionIcon()
                ) == Messages.OK
            }
            if (!install) {
                setup.update(
                    CourseSetupStatus.PLUGINS,
                    SetupStepState.ACTION_NEEDED,
                    message("ui.setup.pluginsDeclined"),
                    SetupAction.INSTALL_PLUGINS
                )
                return true
            }
        }

        val downloadablePluginNodes = pluginsToDownload.map { PluginNode(it) }

        val plugins = application.service<Plugins>()
        val result = plugins.installPlugins(project, downloadablePluginNodes)
        plugins.reportInstallResult(project, result)
        return !result.needsRestart
    }

    /**
     * Records the course's plugins in `.idea/externalDependencies.xml`.
     *
     * The platform's own `CheckProjectRequiredPluginsActivity` reads that file on project open.
     * It notifies the user with install and enable actions.
     */
    private suspend fun registerRequiredPlugins(project: Project, plugins: List<RequiredPlugin>) {
        val requiredIds = (listOf(PluginSettings.PLUGIN_ID) + plugins.map { it.id }).distinct()
        RequiredPluginsFile.write(project, requiredIds)
    }

    private fun displayNamesOf(ids: Set<PluginId>, pluginNames: List<RequiredPlugin>): List<String> {
        val idStrings = ids.map { it.idString }.toSet()
        return pluginNames.filter { it.id in idStrings }.map { it.name }
    }
}
